### PR TITLE
Check for a non-zero denominator in the Ratio/Rational FromCBOR instance

### PR DESCRIPTION
**Pull request by dcoutts**

### CODE REVIEW

**Review by coot — APPROVED** [review]
LGTM

File: binary/src/Cardano/Binary/FromCBOR.hs [hunk 1]
```diff
@@ -184,7 +184,13 @@ instance FromCBOR Int64 where
   fromCBOR = D.decodeInt64
 
 instance (Integral a, FromCBOR a) => FromCBOR (Ratio a) where
-  fromCBOR = (%) <$ enforceSize "Ratio" 2 <*> fromCBOR <*> fromCBOR
+  fromCBOR = do
+    enforceSize "Ratio" 2
+    n <- fromCBOR
+    d <- fromCBOR
+    if d <= 0
```
Comment (redxaxder) [inline binary/src/Cardano/Binary/FromCBOR.hs:191]: Are negative denominators also disallowed?

**Review by redxaxder — COMMENTED** [review]
I believe this should only impact Byron. Is that right?

File: binary/src/Cardano/Binary/FromCBOR.hs (hunk 1 above)
Comment (dcoutts) [reply binary/src/Cardano/Binary/FromCBOR.hs:191]: I _think_ the canonical form for output has negatives in the numerator not the denominator.

**Comment by dcoutts** [comment]
> I believe this should only impact Byron. Is that right?

Oh yes you're right. I didn't notice that `UnitInterval` is using `ratioFromCBOR`, which needs the same zero denominator check.

File: binary/src/Cardano/Binary/FromCBOR.hs (hunk 1 above)
Comment (dcoutts) [reply binary/src/Cardano/Binary/FromCBOR.hs:191]:
```
> 1 % (-1)
(-1) % 1
```

**Pull request merged**